### ISSUE TITLE
Invalid read in inference.jl bootstrap with `GC_VERIFY`

### ISSUE THREAD

**Issue by garrison** [JuliaLang/julia#11455]
I set `CFLAGS = -DMEMDEBUG -DMEMDEBUG2 -DGC_VERIFY` in `Make.user`.  Right now I am also using `LLVM_VER = 3.6.1`, but I doubt this bug is specific to that.  Then,

    $ valgrind -q --smc-check=all-non-file --suppressions=$PWD/contrib/valgrind-julia.supp --trace-children=yes make

[returns numerous invalid reads](https://gist.github.com/garrison/a6b11c060a60df304e98) in `gc_verify` at the `inference.jl` step.

**Comment by garrison** [comment]
Would it make sense to have a label for the various memory bugs (missing GC roots, things found under valgrind or sanitizers, etc)?

**Comment by ScottPJones** [comment]
:+1:

**Comment by garrison** [comment]
BTW (for people attempting to reproduce this): it helps to enable `ccache`, run `make` once, then `make clean`, and then `make` under valgrind, as above.  Or you can run `make` and kill it during the bootstrap, leaving that as the only thing remaining to run under valgrind.  Otherwise you're obviously running the C compiler under valgrind, which can take forever.

**Comment by carnaval** [comment]
I think 4c7bb25a344 fixes it ? Thanks for doing this by the way :-)

**Comment by carnaval** [comment]
(I'd run valgrind but on my laptop I only get the lineno assert error...)

**Comment by garrison** [comment]
@carnaval Oh interesting, what version of LLVM are you using?  I've been considering submitting upstream to valgrind my patch that turns the `lineno` assertion into a warning instead.

**Comment by carnaval** [comment]
I use trunk, that's where the fun's at

**Comment by garrison** [comment]
Here's my [patch to valgrind](https://gist.github.com/garrison/e5ad01cef641c916ecd9), by the way.

**Comment by carnaval** [comment]
I don't think I'll have time to build valgrind with the patch today. If it's easy for you could you confirm that the invalid read does not happen anymore on master ? Otherwise I'll do it tomorrow.

**Comment by garrison** [comment]
It's running, I haven't seen any errors yet.

**Comment by carnaval** [comment]
Thanks. I'll leave it to you to close this when you're convinced it is gone :-)

**Comment by carnaval** [comment]
@garrison I will assume it went well. Feel free to reopen.

**Comment by garrison** [comment]
It's still on `inference.jl`, a day later.  I assume the precise issue I saw has been fixed, though.

**Comment by carnaval** [comment]
Heh, didn't realize it was that slow. memdebug and verify are already heavy, I can only imagine how it goes under valgrind

**Comment by garrison** [comment]
Turns out, things *actually* freeze at that stage of the bootstrap process under valgrind (or under nulgrind).  I've been able to reproduce this with LLVM `3.5.1`, `3.6.1`, and current `svn`.  LLVM `3.3` does not seem to freeze.

Here's the commandline I am running:

    $ valgrind --smc-check=all --trace-children=yes -q --tool=none make

Nothing special is required in `Make.user` -- it seems to happen whether `MEMDEBUG` is defined or not.
